File: nodes/core/market/filters/atr_filter_node.py
import logging

from core.types_registry import IndicatorResult, IndicatorType, IndicatorValue, OHLCVBar
from nodes.core.market.filters.base.base_indicator_filter_node import BaseIndicatorFilter
from services.indicator_calculators.atr_calculator import calculate_atr

logger = logging.getLogger(__name__)


class ATRFilter(BaseIndicatorFilter):
    """
    Filters assets based on ATR (Average True Range) values.
    """

    default_params = {
        "min_atr": 0.0,
        "window": 14,
    }

    params_meta = [
        {"name": "min_atr", "type": "number", "default": 0.0, "min": 0.0, "step": 0.1},
        {"name": "window", "type": "number", "default": 14, "min": 1, "step": 1},
    ]

    def _validate_indicator_params(self):
        min_atr = self.params.get("min_atr", 0.0)
        window = self.params.get("window", 14)
        if not isinstance(min_atr, (int, float)) or min_atr < 0:
            raise ValueError("Minimum ATR cannot be negative")
        if not isinstance(window, (int, float)) or window <= 0:
            raise ValueError("Window must be positive")

    def _calculate_indicator(self, ohlcv_data: list[OHLCVBar]) -> IndicatorResult:
        """Calculate ATR and return as IndicatorResult."""
        if not ohlcv_data:
            return IndicatorResult(
                indicator_type=IndicatorType.ATR,
                timestamp=0,
                values=IndicatorValue(single=0.0),
                params=self.params,
                error="No data",
            )

        window = self.params.get("window", 14)
        if not isinstance(window, (int, float)):
            window = 14
        window = int(window)

        if len(ohlcv_data) < window:
            return IndicatorResult(
                indicator_type=IndicatorType.ATR,
                timestamp=ohlcv_data[-1]["timestamp"],
                values=IndicatorValue(single=0.0),
                params=self.params,
                error="Insufficient data",
            )

        # Extract lists for the calculator
        highs = [bar["high"] for bar in ohlcv_data]
        lows = [bar["low"] for bar in ohlcv_data]
        closes = [bar["close"] for bar in ohlcv_data]

        # Use the calculator - returns full time series
        result = calculate_atr(highs, lows, closes, length=window)
        atr_series = result.get("atr", [])

        # Get the last value from the series
        if atr_series and len(atr_series) > 0:
            latest_atr = atr_series[-1]
        else:
            latest_atr = None

        if latest_atr is None:
            latest_atr = 0.0

        return IndicatorResult(
            indicator_type=IndicatorType.ATR,
            timestamp=ohlcv_data[-1]["timestamp"],
            values=IndicatorValue(single=latest_atr),
            params=self.params,
        )

    def _should_pass_filter(self, indicator_result: IndicatorResult) -> bool:
        """Pass filter if ATR is above minimum threshold."""
        if indicator_result.error or not hasattr(indicator_result.values, "single"):
            return False

        latest_atr = indicator_result.values.single
        min_atr = self.params.get("min_atr", 0.0)
        if not isinstance(min_atr, (int, float)):
            min_atr = 0.0

        return latest_atr >= min_atr

File: nodes/market/filters/atr-filter-node.ts
// src/nodes/core/market/filters/atr-filter-node.ts

import { BaseIndicatorFilter } from './base/base-indicator-filter-node';
import type { NodeDefinition } from '@fig-node/core';
import { IndicatorType, createIndicatorResult, createIndicatorValue, type OHLCVBar, type IndicatorResult } from '../types';
import { calculateAtr } from '../calculators/atr-calculator';

/**
 * Filters assets based on ATR (Average True Range) values.
 */
export class ATRFilter extends BaseIndicatorFilter {
  static override definition: NodeDefinition = {
    ...BaseIndicatorFilter.definition,
    defaults: {
      min_atr: 0.0,
      window: 14,
    },
    params: [
      { name: 'min_atr', type: 'number', default: 0.0, min: 0.0, step: 0.1 },
      { name: 'window', type: 'number', default: 14, min: 1, step: 1 },
    ],
    ui: {
      resultDisplay: 'none',
      actions: [
        { id: 'previewFiltered', label: 'Preview Filtered', icon: '👁' },
        { id: 'copySummary', label: 'Copy Summary', icon: '📋' },
      ],
    },
  };

  protected override validateIndicatorParams(): void {
    const minAtr = this.params.min_atr ?? 0.0;
    const window = this.params.window ?? 14;

    if (typeof minAtr !== 'number' || minAtr < 0) {
      throw new Error('Minimum ATR cannot be negative');
    }
    if (typeof window !== 'number' || window <= 0) {
      throw new Error('Window must be positive');
    }
  }

  protected calculateIndicator(ohlcvData: OHLCVBar[]): IndicatorResult {
    if (!ohlcvData || ohlcvData.length === 0) {
      return createIndicatorResult({
        indicatorType: IndicatorType.ATR,
        timestamp: 0,
        values: createIndicatorValue({ single: 0.0 }),
        params: this.params,
        error: 'No data',
      });
    }

    const windowParam = this.params.window;
    const window = Math.floor(typeof windowParam === 'number' ? windowParam : 14);

    if (ohlcvData.length < window) {
      return createIndicatorResult({
        indicatorType: IndicatorType.ATR,
        timestamp: ohlcvData[ohlcvData.length - 1].timestamp,
        values: createIndicatorValue({ single: 0.0 }),
        params: this.params,
        error: 'Insufficient data',
      });
    }

    // Extract lists for the calculator
    const highs = ohlcvData.map((bar) => bar.high);
    const lows = ohlcvData.map((bar) => bar.low);
    const closes = ohlcvData.map((bar) => bar.close);

    // Use the calculator - returns full time series
    const result = calculateAtr(highs, lows, closes, window);
    const atrSeries = result.atr;

    // Get the last value from the series
    let latestAtr: number | null = null;
    if (atrSeries && atrSeries.length > 0) {
      latestAtr = atrSeries[atrSeries.length - 1];
    }

    if (latestAtr === null) {
      latestAtr = 0.0;
    }

    return createIndicatorResult({
      indicatorType: IndicatorType.ATR,
      timestamp: ohlcvData[ohlcvData.length - 1].timestamp,
      values: createIndicatorValue({ single: latestAtr }),
      params: this.params,
    });
  }

  protected shouldPassFilter(indicatorResult: IndicatorResult): boolean {
    if (indicatorResult.error || indicatorResult.values.single === undefined) {
      return false;
    }

    const latestAtr = indicatorResult.values.single;
    const minAtrParam = this.params.min_atr;
    const minAtr = typeof minAtrParam === 'number' ? minAtrParam : 0.0;

    return latestAtr >= minAtr;
  }
}
